### ISSUE TITLE
Using AdaBoost<DecisionTree>, classification result is all false?

### ISSUE THREAD

**Issue by wup11111** [accord-net/framework#1725]
**What would you like to submit?** (put an 'x' inside the bracket that applies)

  - [x] question
  - [ ] bug report
  - [ ] feature request

**Issue description**

Hi Cesar,
@cesarsouza

I can't predict the correct category label using the Decide method of the AdaBoost algorithm. The prediction result of the test set is all 0 (false). I would like to ask if your AdaBoost classifier supports multi-label classification problems? Since the output is bool type, I tried to write the OvR classification integration myself and converted the training data into two categories (0 and 1). I was surprised to find that the results of all classifiers are false. I exchange the positive and negative (0 and 1) of the training set and the results are all false! Can you tell me how to correctly call the AdaBoost algorithm to solve multi-label classification problems?

I hope you can reply to me, thank you very much!!!

example one:
                                            / / Create multiple learning algorithms for the machine
                                            Var teacher_7_train = new AdaBoost<DecisionTree>()
                                            {
                                                // Here we can specify how each regression should be learned:
                                                Learner = (param) => new C45Learning()
                                                {
                                                    //1. Set the number of times a single variable can be integrated into the decision process. In the original ID3 algorithm, each decision path of a variable can only be connected once (path from root to leaf).
                                                    // If set to zero, a single variable can participate as many times as needed. The default value is 1 and the parameter range is set to [1,200].
                                                    //Join = 100,
                                                    Join = (int)X[num_c][0],

                                                    //2. Set the maximum height allowed when learning the tree. If set to zero, the tree can have any length. The default value is 0 and the parameter range is set to [1,200].
                                                    //MaxHeight = 100,
                                                    MaxHeight = (int)X[num_c][1],

                                                    //3. Set the maximum number of variables that can enter the tree. A value of zero means there is no limit. The default value is 0 and the parameter range is set to [1,200].
                                                    //MaxVariables = 100,
                                                    MaxVariables = (int)X[num_c][2],

                                                },

                                                //——! ! - Training until: -! ! ——
                                                //4. Set the maximum number of iterations that the learning algorithm performs. The parameter range is set to [1,200].
                                                //MaxIterations = 50,
                                                MaxIterations = (int)X[num_c][3],

                                                // Set the error limit before learning stops. The default is 0.5.
                                                Threshold = 0.5,

                                                / / Set the relative tolerance used to detect the convergence of the learning algorithm.
                                                Tolerance = 0.01,

                                            };

                                             // Use the teacher to learn multi-label AdaBoost, learn the enhanced classifier with decision tree algorithm
                                             //Boost<DecisionTree> adaboost_7_train.NumberOfClasses
                                             Var adaboost_7_train = teacher_7_train.Learn(inputs_7_train, outputs_7_train);

                                             / / New AdaBoost judgment array to predict user input data - get predictions for each class of input data
                                             Int[] predicted_7_test = adaboost_7_train.ToMulticlass().Decide(inputs_7_test);

example two:
                                            //——! ! - OvR classification integration (to train num_class(num_class) classifiers) -! ! ——
                                            //——! ! ——##——Each test sample counts num_class classifier results. If more than one class is positive, it is determined according to the prediction confidence—##————! ! ——
                                            / / Record the AdaBoost classifier results of each test sample, a total of n test data
                                            Int[] predicted_7_test = new int[n];

                                            / / Create num_class classifier training results
                                            Boost<DecisionTree>[] adaboost_7_train = new Boost<DecisionTree>[num_class];
                                            / / Create num_class classifier for discriminating
                                            AdaBoost<DecisionTree>[] teacher_7_train = new AdaBoost<DecisionTree>[num_class];
                                            / / Initialize the num_class AdaBoost classifier in turn
                                            #region num_class classifiers have the same four parameters
                                            For (int i = 0; i < num_class; i )
                                            {
                                                / / Create multiple learning algorithms for the machine
                                                Teacher_7_train[i] = new AdaBoost<DecisionTree>()
                                                {
                                                    // Here we can specify how each regression should be learned:
                                                    Learner = (param) => new C45Learning()
                                                    {
                                                        //1. Set the number of times a single variable can be integrated into the decision process. In the original ID3 algorithm, each decision path of a variable can only be connected once (path from root to leaf).
                                                        // If set to zero, a single variable can participate as many times as needed. The default value is 1 and the parameter range is set to [1,200].
                                                        //Join = 100,
                                                        Join = (int)X[num_c][0],

                                                        //2. Set the maximum height allowed when learning the tree. If set to zero, the tree can have any length. The default value is 0 and the parameter range is set to [1,200].
                                                        //MaxHeight = 100,
                                                        MaxHeight = (int)X[num_c][1],

                                                        //3. Set the maximum number of variables that can enter the tree. A value of zero means there is no limit. The default value is 0 and the parameter range is set to [1,200].
                                                        //MaxVariables = 100,
                                                        MaxVariables = (int)X[num_c][2],

                                                    },

                                                    //——! ! - Training until: -! ! ——
                                                    //4. Set the maximum number of iterations that the learning algorithm performs. The parameter range is set to [1,200].
                                                    //MaxIterations = 50,
                                                    MaxIterations = (int)X[num_c][3],

                                                    // Set the error limit before learning stops. The default is 0.5.
                                                    Threshold = 0.5,

                                                    / / Set the relative tolerance used to detect the convergence of the learning algorithm.
                                                    Tolerance = 0.01,

                                                };

                                                //——! ! - Integrated learning does not exist in parallel selection -! ! ——
                                                //The following lines are only used to ensure reproducible results. Please remove it for full parallelization
                                                //teacher_7_train.ParallelOptions.MaxDegreeOfParallelism = 1; //(delete, comment or change this line to enable full parallelism)

                                            }
                                            #endregion
                                            / / Training num_class classifiers in turn
                                            For (int i = 0; i < num_class; i )
                                            {
                                                / / New temporary two types of training set category
                                                Int[] tem_outputs_7_train = new int[m - n];

                                                / / The training set is divided into two classes from num_class class
                                                For (int j = 0; j < (m - n); j )
                                                {
                                                    //The ith class is set to positive (1), not the ith class is set to negative (0)
                                                    If (outputs_7_train[j] == i)
                                                    {
                                                        Tem_outputs_7_train[j] = 0;//正类
                                                    }
                                                    Else
                                                    {
                                                        Tem_outputs_7_train[j] = 1;//negative
                                                    }
                                                }

                                                / / Training the classifier, using the teacher to learn multi-label AdaBoost, using the decision tree algorithm to learn the enhanced classifier
                                                //Boost<DecisionTree> adaboost_7_train.NumberOfClasses
                                                Adaboost_7_train[i] = teacher_7_train[i].Learn(inputs_7_train, tem_outputs_7_train);

                                            }

                                            / / Determine the num_class classifiers for n test samples in turn
                                            For (int i = 0; i < n; i++)
                                            {
                                                / / Statistics of the positive and negative of each classifier
                                                Bool[] each_classifier = new bool[num_class];//Data contains num_class classes
                                                / / Statistics of each classifier's prediction confidence
                                                Double[] each_confidence = new double[num_class];//The data contains num_class classes
                                                //Statistics have several classifiers true
                                                Int true_classifier = 0;
                                                / / Record the number with the classifier is true
                                                Int[] code_classifier = new int[num_class];

                                                / / Calculate the positive and negative of each classifier
                                                For (int j = 0; j < num_class; j++)
                                                {
                                                    / / Get the positive and negative classification results of the sample
                                                    Each_classifier[j] = adaboost_7_train[j].Decide(inputs_7_test[i]);
                                                    ////Get the decision tree
                                                    //each_confidence[j] = adaboost_7_train[j].Models[0].Model.;
                                                }

                                                / / Check the results of several classifiers are positive
                                                For (int j = 0; j < num_class; j++)
                                                {
                                                    If (each_classifier[j] == true)
                                                    {
                                                        Code_classifier[true_classifier] = j;//record number
                                                        True_classifier = true_classifier + 1; / / number of records
                                                    }
                                                }

                                                / / Only one is true, directly specified as the class
                                                If (true_classifier == 1)
                                                {
                                                    Predicted_7_test[i] = code_classifier[0];
                                                }
                                                // When all is false, set the category directly to -1
                                                Else if (true_classifier == 0)
                                                {
                                                    Predicted_7_test[i] = -1;
                                                }
                                                / / When there are multiple true, compare the confidence of the two classifiers
                                                Else if (true_classifier > 1)
                                                {
                                                    Predicted_7_test[i] = 100;
                                                }
                                            }

Best regards.
Wu Peng
